packages/pyright-internal/src/analyzer/sourceFile.ts

```diff
@@ -77,6 +78,60 @@ export enum IPythonMode {
     CellDocs,
 }
 
+/**
+ * gets the content of a file. if it's a notebook, the whole notebook json is returned as it appears in the file
+ */
+const getFileContent = (fileSystem: FileSystem, uri: Uri, console: ConsoleInterface): string | undefined => {
+    try {
+        // Check the file's length before attempting to read its full contents.
+        const fileStat = fileSystem.statSync(uri);
+        if (fileStat.size > maxSourceFileSize) {
+            console.error(
+                `File length of "${uri}" is ${fileStat.size} ` +
+                    `which exceeds the maximum supported file size of ${maxSourceFileSize}`
+            );
+            throw new Error('File larger than max');
+        }
+
+        return fileSystem.readFileSync(uri, 'utf8');
+    } catch (error) {
+        return undefined;
+    }
+};
+
+/**
+ * if this is a notebook, gets the cells in this file that contain python code.
+ * note that each {@link SourceFile} is an individual cell, but this function returns
+ * all of them
+ * @returns `undefined` if not a notebook
+ */
+export const getIPythonCells = (fileSystem: FileSystem, uri: Uri, console: ConsoleInterface) => {
+    if (!uri.hasExtension('.ipynb')) {
+        return undefined;
+    }
+    const fileContent = getFileContent(fileSystem, uri, console);
+    if (!fileContent) {
+        return undefined;
+    }
+    const parsedNotebook = JSON.parse(fileContent) as INotebookContent;
+    return parsedNotebook.cells.filter(
```

Comment:
filter out non-python notebooks? or are all `.ipynb` files guaranteed to be python ones?

Reply:
Some cells could be markdown. There could be other kernels, but I'd hope that this line would have taken care of that

`{ notebook: '*', language: 'python' },`

Otherwise, there's `metadata.language_info.name` property in the `.ipynb` file that can drive this filter.

Reply:
There are some funky multi-language kernels out there, but pretty rare. By and large, a notebook is one executable language.

There are also cell magics that let one do e.g. inline JavaScript, might be nice to filter those out, but it's a nice-to-have for sure.

Reply:
when running as a language server,  the parsing and filtering of the cells is handled automatically. `{ notebook: '*', language: 'python' }` is for the language server but this code here is only used by the CLI

Reply:
👍 filtering out only code cells as you're doing here is probably good in 99% of cases